eval.ml

```diff
@@ -199,34 +200,46 @@ let reduce conf =
        trace = v :: conf.trace;
        pc = pc';
      }
-  | Invalidate (e, l, xs) ->
+  | Osr (e, v, l, osr) ->
      let b = get_bool (eval conf e) in
-     if b then
+     if not b then
        { conf with
          pc = pc';
        }
      else begin
-       let add env x =
-         match Env.find x conf.env with
-         | exception Not_found -> raise (Unbound_variable x)
-         | v -> Env.add x v env
+       let add (env, heap) = function
+         | OsrConst (x, e) ->
+           (Env.add x (Const (eval conf e)) env,
+            heap)
+         | OsrMut (x, OsrExp e) ->
+           let a = Address.fresh () in
+           (Env.add x (Mut a) env,
+            Heap.add a (Value (eval conf e)) heap)
+         | OsrMut (x, OsrUndef) ->
+           let a = Address.fresh () in
+           (Env.add x (Mut a) env,
+            Heap.add a Undefined conf.heap)
```

Comment:
Having a heap that is never garbage-collected is ok-ish. (In particular there is never any garbage collection at this level.) Note that `drop`-ing a mutable variable properly shrinks the heap, so this problem will go away with strict dropping rules.